funsor/optimizer.py

```diff
@@ -0,0 +1,205 @@
+        # Case 1) Finitary(Finitary) -> Finitary
+        new_operands = []
+        for term in operands:
+            if isinstance(term, Finitary) and term.op == op:
+                new_operands.extend(term.operands)
+            else:
+                new_operands.append(term)
+
+        with interpretation(reflect):
+            return Finitary(op, tuple(new_operands))
+    elif all(isinstance(term, Reduce) for term in operands):  # TODO check distributivity
+        # Case 2) Finitary(Reduce, Reduce) -> Reduce(Finitary(lhs.arg, rhs.arg))
+        new_operands = []
+        new_reduced_vars = set()
+        for term in operands:
+            new_operands.append(term.arg)
+            new_reduced_vars = new_reduced_vars.union(term.reduced_vars)
+
+        with interpretation(reflect):
+            return Reduce(operands[0].op, Finitary(op, tuple(new_operands)), new_reduced_vars)
+    elif all(not isinstance(term, (Reduce, Finitary)) for term in operands):
+        with interpretation(reflect):
+            return Finitary(op, operands)  # nothing to do, reflect
+    else:
+        # Note: if we can't rewrite all operands in the finitary, fail for now
+        # A more sophisticated strategy is to apply this rule recursively
+        # Alternatively, we could do this rewrite on Binary ops instead of Finitary
+        raise NotImplementedError("TODO(eb8680) handle mixed case")
+
+
+@deoptimize.register(Reduce, object, Reduce, frozenset)
+def deoptimize_reduce(op, arg, reduced_vars):
+    """
+    Rewrite to the largest possible Reduce(Finitary) by combining Reduces
+    Assumes that all input Reduce/Finitary ops have been rewritten
+    """
+    if arg.op == op:
+        # Reduce(Reduce) -> Reduce
+        new_reduced_vars = reduced_vars.union(arg.reduced_vars)
+        return Reduce(op, arg.arg, new_reduced_vars)
+    return None
+
+
+def optimize(cls, *args):
+    result = _optimize(cls, *args)
+    if result is None:
+        result = reflect(cls, *args)
+    return result
+
+
+_optimize = KeyedRegistry(default=lambda *args: None)
+optimize.register = _optimize.register
+
+
+# TODO set a better value for this
+REAL_SIZE = 3  # the "size" of a real-valued dimension passed to the path optimizer
+
+
+@optimize.register(Reduce, object, Finitary, frozenset)
+def optimize_reduction(op, arg, reduced_vars):
+    r"""
+    Recursively convert large Reduce(Finitary) ops to many smaller versions
+    by reordering execution with a modified opt_einsum optimizer
+    """
+    # build opt_einsum optimizer IR
+    inputs = []
+    size_dict = {}
+    for operand in arg.operands:
+        inputs.append(frozenset(d for d in operand.inputs.keys()))
+        size_dict.update({k: (REAL_SIZE if v.dtype == 'real' else v.dtype)
+                          for k, v in operand.inputs.items()})
+    outputs = frozenset().union(*inputs) - reduced_vars
+
+    # optimize path with greedy opt_einsum optimizer
+    path = greedy(inputs, outputs, size_dict)
```

Comment:
FYI we should switch to the new [auto strategy[(https://github.com/dgasmith/opt_einsum/commit/f9afb0b4ba440a247ff55312e8cc20ec38752b71) once it is released.